sdks/python/apache_beam/runners/interactive/pipeline_instrument.py

```diff
@@ -95,17 +119,129 @@ def instrumented_pipeline_proto(self):
+    pcollection_ids = [pc for t in required_transforms.values()
+                       for pc in t.outputs.values()]
+    required_pcollections = {pc_id: pcollections[pc_id]
+                             for pc_id in pcollection_ids}
+
+    # Cache all the PCollection coders.
+    coder_ids = [pc.coder_id for pc in required_pcollections.values()]
+    required_coders = {c_id: coders[c_id] for c_id in coder_ids}
+
+    # Cache all the windowing strategy ids.
+    windowing_strategies_ids = [pc.windowing_strategy_id
+                                for pc in required_pcollections.values()]
+    required_windowing_strategies = {ws_id: windowing_strategies[ws_id]
+                                     for ws_id in windowing_strategies_ids}
+
+    subtransforms = {}
+    subpcollections = {}
+    subcoders = {}
+    subwindowing_strategies = {}
+
+    # Recursively go through all the subtransforms and add their components.
+    for transform_id, transform in required_transforms.items():
+      if transform_id in pipeline_proto.root_transform_ids:
+        continue
+      (t, pc, c, ws) = self._required_components(pipeline_proto,
+                                                 transform.subtransforms)
+      subtransforms.update(t)
+      subpcollections.update(pc)
+      subcoders.update(c)
+      subwindowing_strategies.update(ws)
+
+    # Now we got all the components and their subcomponents, so return the
+    # complete collection.
+    required_transforms.update(subtransforms)
+    required_pcollections.update(subpcollections)
+    required_coders.update(subcoders)
+    required_windowing_strategies.update(subwindowing_strategies)
+
+    return (required_transforms, required_pcollections, required_coders,
+            required_windowing_strategies)
+
+  def background_caching_pipeline_proto(self):
+    """Returns the background caching pipeline.
+
+    This method creates a background caching pipeline from the original user
+    pipeline by: adding writes to cache from each unbounded source (done in the
+    instrument method), and cutting out all components (transform, PCollections,
+    coders, windowing strategies) that are not the unbounded sources or writes
+    to cache (or subtransforms thereof).
+    """
+    # Create the pipeline_proto to read all the components from. It will later
+    # create a new pipeline proto from the cut out components.
+    pipeline_proto = self._background_caching_pipeline.to_runner_api(
+        return_context=False, use_fake_coders=True)
+
+    # Get all the sources we want to cache.
+    sources = unbounded_sources(self._background_caching_pipeline)
+
+    # Get all the root transforms. The caching transforms will be subtransforms
+    # of one of these roots.
+    roots = pipeline_proto.root_transform_ids
+
+    # Get the transform IDs of the caching transforms.
+    transforms = pipeline_proto.components.transforms
+    caching_transform_ids = [t_id for root in roots
```

Comment:
Why do we need to look for WRITE_CACHE transforms? The original user pipeline shouldn't contain them, right?

Reply:
Thanks for the clarifying question. I reworded the comment here and in the method comment header.

I think my wording in the method comment and here were insufficient. This method is making a background caching pipeline proto from the _background_caching_pipeline. The writes to cache are added to the _background_caching_pipeline in the instrument() method. This is to keep idempotency in the background_caching_pipeline_proto() method (multiple calls won't add multiple writes to cache).